example/bwmf/bwmf_task.go

```diff
@@ -74,12 +74,150 @@ type bwmfTask struct {
 	childrenReady map[uint64]bool
 
 	dShard, tShard Shard
-	d, t           []*bwmfData
+	d, t           Shard
+
+	// parameters for projected gradient methods
+	sigma, alpha, beta    float64
+	maxIterInSubProblem   int
+	maxIterInFindingAlpha int
 }
 
 // These two function carry out actual optimization.
-func (t *bwmfTask) updateDShard() {}
-func (t *bwmfTask) updateTShard() {}
+func (t *bwmfTask) updateDShard() {
+	var newDShard Shard = t.dShard
+	var ok bool
+	for iter := 0; iter < t.maxIterInSubProblem; iter++ {
+		newDShard, ok = t.solveSubproblem(&t.rowShard, &newDShard, &t.t)
+		// TODO(baigang): dealing with ok==false
+		// TODO(baigang): logging
+	}
+	t.dShard = newDShard
+}
+func (t *bwmfTask) updateTShard() {
+	var newTShard Shard = t.tShard
+	var ok bool
+	for iter := 0; iter < t.maxIterInSubProblem; iter++ {
+		newTShard, ok = t.solveSubproblem(&t.columnShard, &t.tShard, &t.d)
+		// TODO(baigang): dealing with ok==false
+		// TODO(baigang): logging
+	}
+	t.tShard = newTShard
+}
+
+// calculates $matrix^T * matrix$
+func matrixSelfTransMult(matrix *Shard) Shard {
```

Comment:
Are we making assumption that we can compute this in memory for the entire matrix?